de: port to new Zig JSON API

LordMZTE opened the pull request:
This PR updates to the new Zig JSON API

Noteworthy changes and stuff to watch out for:

- `fromReader` and `fromReaderWith` functions have been added
- `Deserializer` now takes an extra generic argument for a reader
- In `de.free`, I'm just passing it some random reader from std (wonk)
- Allocators are now mandatory due to `std.json.Reader` requiring one.
- Some errors that are no longer in the std API have been renamed, namely `InvalidTopLevelTrailing` and `UnbalancedBraces` are both `SyntaxError` now.
    - The error set in the public API has changed.
- New peek function is being utilized, got rid of one particularly horrible specimen of a hack.
- Some places that used to use `is_integer` to check for ints vs floats now just try parsing as int and return `error.InvalidType` in case of failure. This improves performace, as we now need to manually check the type of a number literal using `std.json.isNumberFormattedLikeAnInteger` (good function naming there, andrew).
- When switching over tokens, numbers and strings require an inline switch because the stack and heap allocated tokens are `[]const u8` and `[]u8` respectively which don't coerce when being used in one switch branch.
- String unescaping is now entirely handled by `std.json`.
    - Thus, `unescapeString` has been removed.
- The allocator stored in the deserializer can be conveniently retrieved using `self.tokens.scanner.stack.bytes.allocator` (or `self.getAllocator()`)
- `Union` needs to return allocated keys somtimes. This has been worked around for now by having a buffer field which the value is copied into before being returned.

LordMZTE committed:
de: port to new Zig JSON API

phan-do commented on src/de/deserializer.zig:
```diff
@@ -2,36 +2,64 @@ const concepts = @import("concepts");
+            }
+        }
+
+        /// A small helper to retrieve this struct's stored allocator.
+        pub inline fn getAllocator(self: *const Self) std.mem.Allocator {
+            return self.tokens.scanner.stack.bytes.allocator;
+        }
```
Let's remove this and store the allocator passed to `init` ourselves.

The allocator returned here is used by the `BitStack` data structure to track `[]` and `{}` nesting levels; it's not really intended for allocating or freeing parsed values returned by `nextAlloc` like strings or integers.[[1]](https://github.com/ziglang/zig/blob/41502c6aa53a3da31b276c23c4db74db7d04796b/lib/std/json/scanner.zig#L230)

phan-do replied:
Actually, I don't even think we need to store the allocator anywhere nor reference the `BitStack`'s allocator.

All of the deserialize methods take an optional allocator, so it should be fine to just use that, no? Plus, IIRC, the allocator field in `Deserializer` isn't even used anywhere atm, so...

phan-do replied:
_From Discord..._

> about getAllocator, it's currently used in a few functions where I don't have any other allocator:
> - freeToken
> - fromDeserializer

Yeah, we do need to have an allocator lying around for `fromDeserializer`. That's fine.

> having another field would be storing the allocator twice, which is kind of wasteful. you also see the pattern of taking an allocator from a child data structure a lot in std.
> 
> i do get your point about that comment in std's json stating the allocator is only used for tracking nesting, but then again that doesn't really mean we can't use it for anything else.

It's fine if we store the allocator twice; it's just two pointers. We can easily optimize this whenever we want if it turns out to be an issue.

I don't want to have to rely on `BitStack`'s implementation to grab our allocator. For example, if its `bytes` field changes from an `ArrayList` or is removed or whatever, then we'd probably have to change our code as well.

Also, even though `BitStack`'s allocator is exact same as the one passed to `init`, I just think it's confusing and messy to have to reach all the down into `self.tokens.scanner.stack.bytes.allocator` just to get our allocator that was passed to `init`.

> Also, I currently use the allocator passed to the deserialize functions only for stuff that is returned from the function in case it seemed random which one i use

For consistency with what we had before (and in order to be proper... um, Getty citizens I guess?) let's always use the passed-in allocator in the `deserialize` methods (and whatever other functions take them). If `allocator == null`, we can return `error.MissingAllocator`.

The new `allocator` field in `Deserializer` will be for `fromDeserializer` calls.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -2,36 +2,64 @@ const concepts = @import("concepts");
+                else => false,
+            }) {}
+        }
+
```
Let's move these functions below the `getty.Deserializer` implementation code. Generally, I like keeping private functions after wherever they're used in public API.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -55,161 +83,151 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
-            std.json.TokenStream.Error ||
-            error{UnexpectedJsonDepth} || // may be returned while skipping values
+            std.json.Error ||
+            error{ValueTooLong} || // may be returned from nextAlloc
```
```suggestion
            JsonReader.AllocError ||
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -55,161 +83,151 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
+            self: *Self,
+            allocator: ?std.mem.Allocator,
+            visitor: anytype,
+        ) Error!@TypeOf(visitor).Value {
```
Can we make this one line? Breaking it up makes it even harder to read imo.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -2,36 +2,64 @@ const concepts = @import("concepts");
-pub fn Deserializer(comptime user_dbt: anytype) type {
+pub fn Deserializer(comptime user_dbt: anytype, comptime Reader: type) type {
+    const JsonReader = std.json.Reader(1024 * 4, Reader);
     return struct {
```
```suggestion
    const JsonReader = std.json.Reader(1024 * 4, Reader);

    return struct {
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -2,36 +2,64 @@ const concepts = @import("concepts");
+        }
+
+        /// Frees a Token if it is an allocated variant.
+        fn freeToken(self: *const Self, tok: std.json.Token) void {
```
I noticed a lot of `*const Self` in the PR. Is there a performance reason for this over using just a `Self`?

LordMZTE replied:
You're right, Self is better here.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -55,161 +83,151 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
-                    const slice = token.Number.slice(self.tokens.slice, self.tokens.i - 1);
+            const token = try self.tokens.nextAlloc(self.getAllocator(), .alloc_if_needed);
+            defer self.freeToken(token);
+            switch (token) {
```
```suggestion
            defer self.freeToken(token);

            switch (token) {
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -249,137 +267,104 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
                         }
 
-                        return try visitor.visitInt(allocator, De, try std.fmt.parseInt(Value, slice, 10));
+                        return try visitor.visitInt(allocator, De, std.fmt.parseInt(Value, slice, 10) catch return error.InvalidType);
```
Why return `error.InvalidType` here? We still have `std.fmt.ParseIntError` in our error type we pass to getty, so the `try` should still work.

Also, the other `parseInt` and `parseFloat` calls still use `try`.

LordMZTE replied:
The reason for this is because those are in contexts where a ParseIntError should only occur in case the JSON data contains a float, which is the wrong data type. It's true that I missed a few places where this also is the case but it's still using try.

phan-do replied:
`ParseIntError` can also be returned by `std.fmt.parseInt` in the case of overflow (i.e., when the input data contains a number too large to store in `Value`). In which case, `InvalidType` wouldn't be the more accurate error to return.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -249,137 +267,104 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
-                    return try visitor.visitMap(allocator, De, map.mapAccess());
-                }
+            // Don't need nextAlloc here, as data that would need allocation
+            // is always an invalid type.
```
nit: can remove this comment.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -249,137 +267,104 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
-                    var s = StructAccess(Self){ .d = self };
-                    return try visitor.visitMap(allocator, De, s.mapAccess());
-                }
+            if (try self.tokens.next() == .object_begin) {
```
TIL this works now 🎉  Usually, I used to have to wrap `try self.tokens.next()` in parentheses

phan-do commented on src/de/deserializer.zig:
```diff
@@ -249,137 +267,104 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
-                    var u = Union(Self){ .d = self };
-                    return try visitor.visitUnion(allocator, De, u.unionAccess(), u.variantAccess());
-                }
+            switch (try self.tokens.peekNextTokenType()) {
+                .string => {},
+                .object_begin => try self.skipToken(),
+                else => return error.InvalidType,
```
Can we make this the default approach for all the other deserialize methods?

That is, peek first and then skip a token if everything looks okay. That'll make the deserializer peekable and would let us avoid modifying the parser's state in obviously incorrect situations.

LordMZTE replied:
This approach would have a few issues:
- `peekNextTokenType` only returns the type of the next token, not the contents. We'd still have to call next to get the actual data.
- We always error out in the case of an unexpected token, thus the state the parser is left in is irrelevant.
- First peeking and then getting the token afterwards would have large performance implications, because we'd essentially be parsing everything twice.

I think only peeking if necessary makes more sense here, pretty much for the same reason that the best way to check if a file I want to read exists is to just try to open it and handle errors. It's one less operation, doesn't do any harm and yields the same result.

phan-do replied:
> peekNextTokenType only returns the type of the next token, not the contents. We'd still have to call next to get the actual data.

Sorry, I meant peek first and then get the next token with `next`/`nextAlloc` (not skip it).

> First peeking and then getting the token afterwards would have large performance implications, because we'd essentially be parsing everything twice.

I think it'd actually help performance.

AFAIU, `peekNextTokenType` basically just switches on `self.state`, compares it against a char, and then returns a token. That's a lot less work than what `nextAlloc` does (e.g., modifying the stack, fully parsing numbers, allocating integers/strings, etc.).

In other words, peeking first and then calling `next`/`nextAlloc` costs us one extra switch, and in return we get to fail faster in the bad path since we never have to allocate or parse anything if we encounter an unexpected type. Not to mention, our deserializer becomes peekable, which can be useful as described 👇 .

> We always error out in the case of an unexpected token, thus the state the parser is left in is irrelevant.

~~It may be relevant depending on how the deserializer's used.~~

~~For example, a while back, I recall you trying to deserialize into a union by writing your own DB that first called `deserializer.deserializeMap` and then, if that failed, called `deserializer.deserializeString` (or something like that). However, that approach doesn't work if the deserializer isn't peekable, since its state would've been modified when `deserializeMap` failed.~~

~~Making `Deserializer` peekable would enable this kind of use case and would cost us virtually nothing.~~

edit: bad example, since it wouldn't really work as the parser's state eventually progresses so you wouldn't be able to call another deserialize method anyways. Thanks for pointing this out @LordMZTE!

phan-do commented on src/de/deserializer.zig:
```diff
@@ -440,43 +425,17 @@ fn MapAccess(comptime D: type) type {
+                inline .string, .allocated_string => |string| {
+                    var mkd = MapKeyDeserializer(De){ .key = string };
+                    var result = try seed.deserialize(allocator, mkd.deserializer());
+                    return result;
```
```suggestion
                    return try seed.deserialize(allocator, mkd.deserializer());
```

phan-do reviewed (changes requested):
First pass.

Really great work! I'll try to get through more over the weekend (or maybe after this review's resolved).

LordMZTE committed:
implement requested changes

- added allocator field and removed getAllocator
- made function headers single-line
- moved internal functions below deserializer implementation
- added some spacing lines
- updated error set
- use passed allocator in deserialize functions
- improved handling of errors from parseInt
- removed a useless temporary variable

phan-do commented on src/de/deserializer.zig:
```diff
@@ -633,19 +585,10 @@ fn Union(comptime D: type) type {
-
-    try std.json.unescapeValidString(escaped, slice);
-    return escaped;
+/// Like std.fmt.parseInt, but does some error conversions to better fit getty's API
```
`parseInt` isn't public, so we should use a normal comment here.

```suggestion
// Like std.fmt.parseInt, but does some error conversions to better fit getty's API
```

LordMZTE replied:
What's wrong with documenting internal functions? They won't be in the public API does anyways, and it's more fitting. Back when autodoc had internal doc mode, it also would've supported this. (I think there's still something like that)

phan-do replied:
I personally like using `///` for public API stuff and `//` for everything else. Just a personal preference though, we can keep it as `///` here.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -633,19 +585,10 @@ fn Union(comptime D: type) type {
+    return std.fmt.parseInt(T, slice, radix) catch |e| switch (e) {
+        error.InvalidCharacter => error.InvalidType,
+        error.Overflow => error.Overflow,
+    };
```
nit

```suggestion
    return std.fmt.parseInt(T, slice, radix) catch |err| switch (err) {
        error.InvalidCharacter => error.InvalidType,
        error.Overflow => |e| e,
    };
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -598,33 +546,37 @@ fn Union(comptime D: type) type {
 
         fn payloadSeed(self: *Self, allocator: ?std.mem.Allocator, seed: anytype) Error!@TypeOf(seed).Value {
-            if (@TypeOf(seed).Value != void) {
+            if (comptime @TypeOf(seed).Value != void) {
```
`comptime` is redundant here.

```suggestion
            if (@TypeOf(seed).Value != void) {
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -598,33 +546,37 @@ fn Union(comptime D: type) type {
+                    else => return error.SyntaxError,
                 }
 
                 return payload;
```
nit

```suggestion
                return switch (try self.d.tokens.next()) {
                    .object_end => payload,
                    else => return error.SyntaxError,
                };
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -580,6 +524,10 @@ fn Union(comptime D: type) type {
 
+        // TODO: allow returning allocated variants from variantSeed
+        /// A place to temporarily store heap-allocated variants.
+        variant_buf: [512]u8 = undefined,
```
I think we can make this a `[]u8` and just assign to it the allocated string directly. It's allocated, so it's not going anywhere after we return from `variantSeed`.

phan-do replied:
Also, we'd probably need to rename it to something like `allocated_variant` or something. I'll leave it up to you.

LordMZTE replied:
That would be a memory leak. The reason I copy it to the buffer is in order to be able to free it afterwards (done by `defer self.d.freeToken` here. If I were to return it directly, there would be no way to free the token at all, and since the buffer here is a part of the struct, it's on the stack.

phan-do replied:
Oh, I see now. I thought you were freeing it elsewhere but no it's freed in `variantSeed` itself.

Makes sense!

phan-do commented on src/de/deserializer.zig:
```diff
@@ -546,24 +489,25 @@ fn StructAccess(comptime D: type) type {
-                    .String => |t| {
-                        const slice = t.slice(self.d.tokens.slice, self.d.tokens.i - 1);
+            // We use the allocator passed here rather than the allocator in the
+            // deserializer since we may need to return allocated strings.
```
We can remove this since we use the passed-in allocator everywhere now.
```suggestion
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -546,24 +489,25 @@ fn StructAccess(comptime D: type) type {
-                        const slice = t.slice(self.d.tokens.slice, self.d.tokens.i - 1);
+            // We use the allocator passed here rather than the allocator in the
+            // deserializer since we may need to return allocated strings.
+            const token = try self.d.tokens.nextAlloc(allocator.?, .alloc_if_needed);
```
need a `if (allocator == null)` check before this line.

phan-do commented on src/de/deserializer.zig:
```diff
@@ -440,43 +418,16 @@ fn MapAccess(comptime D: type) type {
-                    else => {},
-                }
+            const token = try self.d.tokens.nextAlloc(allocator orelse return error.MissingAllocator, .alloc_if_needed);
+            defer self.d.freeToken(token);
+            switch (token) {
```
nit
```suggestion
            defer self.d.freeToken(token);
            
            switch (token) {
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -249,137 +237,127 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
+        /// Frees a Token if it is an allocated variant.
+        fn freeToken(self: Self, tok: std.json.Token) void {
+            switch (tok) {
+                inline .allocated_number,
+                .allocated_string,
+                => |slice| self.allocator.free(slice),
```
nit
```suggestion
                inline .allocated_number, .allocated_string => |s| self.allocator.free(s),
```

phan-do commented on src/de/deserializer.zig:
```diff
@@ -66,150 +67,137 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
+                    return try visitor.visitBool(allocator, De, token == .true);
+                },
+                inline .number, .allocated_number => |slice| {
+                    const is_integer = std.json.isNumberFormattedLikeAnInteger(slice);
```
nit

We can rearrange this section so that we only call `isNumberFormattedLikeAnInteger` when we actually need it.

To do so, we can move the `if (visitor_info == .Int)` block up to the top of this branch. Then, the `if (visitor_info == .Enum)` block will be below. Then, we have the float section with `if (std.json.isNumberFormattedLikeAnInteger)`. And then after that the `return switch(slice[0])` for integers with no type hints.

phan-do commented on src/de.zig:
```diff
@@ -23,39 +23,55 @@ pub const de = struct {
     return try fromSliceWith(allocator, T, slice, null);
 }
 
+pub fn fromReader(allocator: std.mem.Allocator, comptime T: type, reader: anytype) !T {
+    return try fromReaderWith(allocator, T, reader, null);
+}
+
+pub fn fromReaderWith(allocator: std.mem.Allocator, comptime T: type, reader: anytype, comptime user_dbt: anytype) !T {
+    var d = Deserializer(user_dbt, @TypeOf(reader)).init(allocator, reader);
+    defer d.deinit();
+    return fromDeserializer(T, &d);
+}
```
Can you move these function above `fromSliceWith` and swap their positions?

Reading from the top of the file, we should get the innermost function `fromDeserializer`, then `fromReaderWith`, which uses `fromDeserializer`, then `fromReader`, which uses `fromReaderWith`, and so on.

phan-do reviewed (changes requested):
Looks good! We can merge after this review.

It's mostly nits. Thanks for all the great work!

LordMZTE committed:
implement requested changes

- reposition `fromReader` and `fromReaderWith`
- optimize use of `std.json.isNumberFormattedLikeAnInteger`
- change some formatting
- do null checks for allocator
- remove pointless use of `comptime`
